scripts/validate_docstrings.py

```diff
@@ -45,86 +45,188 @@
+    previous_line = current_section = current_subsection = ''
+    position = None
+    for line in api_doc_fd:
+        line = line.strip()
+        if len(line) == len(previous_line):
+            if set(line) == set('-'):
+                current_section = previous_line
+                continue
+            if set(line) == set('~'):
+                current_subsection = previous_line
+                continue
 
-    for part in func_parts:
-        obj = getattr(obj, part)
-    return obj
+        if line.startswith('.. currentmodule::'):
+            current_module = line.replace('.. currentmodule::', '').strip()
+            continue
 
+        if line == '.. autosummary::':
+            position = 'autosummary'
+            continue
 
-def _to_original_callable(obj):
-    while True:
-        if inspect.isfunction(obj) or inspect.isclass(obj):
-            f = inspect.getfile(obj)
-            if f.startswith('<') and f.endswith('>'):
-                return None
-            return obj
-        if inspect.ismethod(obj):
-            obj = obj.__func__
-        elif isinstance(obj, functools.partial):
-            obj = obj.func
-        elif isinstance(obj, property):
-            obj = obj.fget
-        else:
-            return None
+        if position == 'autosummary':
+            if line == '':
+                position = 'items'
+                continue
 
+        if position == 'items':
+            if line == '':
+                position = None
+                continue
+            item = line.strip()
+            func = importlib.import_module(current_module)
+            for part in item.split('.'):
+                func = getattr(func, part)
 
-def _output_header(title, width=80, char='#'):
-    full_line = char * width
-    side_len = (width - len(title) - 2) // 2
-    adj = '' if len(title) % 2 == 0 else ' '
-    title_line = '{side} {title}{adj} {side}'.format(side=char * side_len,
-                                                     title=title,
-                                                     adj=adj)
+            yield ('.'.join([current_module, item]), func,
+                   current_section, current_subsection)
 
-    return '\n{full_line}\n{title_line}\n{full_line}\n\n'.format(
-        full_line=full_line, title_line=title_line)
+        previous_line = line
 
 
 class Docstring(object):
-    def __init__(self, method_name, method_obj):
-        self.method_name = method_name
-        self.method_obj = method_obj
-        self.raw_doc = method_obj.__doc__ or ''
-        self.clean_doc = pydoc.getdoc(self.method_obj)
+    def __init__(self, name):
+        self.name = name
+        obj = self._load_obj(name)
+        self.obj = obj
+        self.code_obj = self._to_original_callable(obj)
+        self.raw_doc = obj.__doc__ or ''
+        self.clean_doc = pydoc.getdoc(obj)
         self.doc = NumpyDocString(self.clean_doc)
 
     def __len__(self):
         return len(self.raw_doc)
 
+    @staticmethod
```

Comment:
All of the calls within the code base are using an instance of the class (save the example in the docstring). With that being the case would prefer to remove the decorator unless there's actually a use case for it; otherwise just adds fluff

Reply:
Not a problem at all to make it a regular method, I leave often static methods as regular methods, as it makes almost no difference.

But I don't understand your comment about using an instance of the class. Do you mean that it's not something that we usually use?

The way I see it is that methods of classes in general access the attributes and other methods of the class. As they are encapsulate a set of data and code that somehow works together. Functions in the other hand are somehow independent. In few cases, there are functions that while they do not access anything else from the class, they make more sense inside the class than outside. This is the case here, as `_load_obj` is only used by other methods of the class, and living outside at the same level of the class `Docstring`, IMO makes it look more important in the code than what really is, as conceptually it makes more sense at the same level of the other methods.

Assuming that as a method of the class is a better decision, having it as an staticmethod makes very quick to understand that it's a function inside of a class, and not a method that can access other methods or attributes. I personally find this information very valuable when reading code. Normal methods can access anything in the class, adding an extra complexity in the way that you can't see them as pure input->output units, but that can instead use somehow global variables. A static method doesn't have this complexity, and I know I can expect it to use only whatever parameter it receives, return a result, and not modify anything else. So, it gives a lot of information without having to read its code. The cost is just having the text `@staticmethod` instead of `self`. I think the few characters of difference is a very good price for the amount of information it provides. And I don't think that not using them often in the pandas code itself is a good reason for not using it here. May be we should start using them more there.

Reply:
\+1 on explicitly using staticmethod, if it is actually one

Reply:
My point is that I don't see the benefit of just adding things like this without any accompanying test regardless of how trivial the change may seem

Reply:
I'm a bit lost on what's the connection between the staticmethod and the testing here.

The benefit of this PR is that makes the code more organized and better. And it implements a way to extract detailed information of the errors in the docstrings. So I can quickly generate the list of functions that fail a specific validation, fix the problem in all them, and start checking in the CI those validations. So the errors don't happen again, and I don't need to waste time on manually reviewing them in all the docstrings PRs.

A huge benefit for me. :)  So would be great to have this merged. Is there any blocker, or can I merge on green?

Reply:
I've blocked long enough so if others are fine with it majority rules :-) 

Just generally trying to minimize PRs and keep them directed towards one clear goal at a time, hence the pushback on this one

Reply:
hehe, I'm trying to maximize PRs, so hopefully we'll end up in a good equilibrium. ;) After this is merged I'll start creating issues for the points in #20298, so we have better validation of the docstrings.

@jorisvandenbossche happy to merge this?